pkg/sql/sem/tree/eval.go

```diff
@@ -1852,70 +1986,74 @@ func (o cmpOpOverload) LookupImpl(left, right *types.T) (*CmpOp, bool) {
 }
 
 func makeCmpOpOverload(
-	fn func(ctx *EvalContext, left, right Datum) (Datum, error), a, b *types.T, nullableArgs bool,
+	fn func(ctx *EvalContext, left, right Datum) (Datum, error),
+	a, b *types.T,
+	nullableArgs bool,
+	v Volatility,
 ) *CmpOp {
 	return &CmpOp{
 		LeftType:     a,
 		RightType:    b,
 		Fn:           fn,
 		NullableArgs: nullableArgs,
+		Volatility:   v,
 	}
 }
 
-func makeEqFn(a, b *types.T) *CmpOp {
-	return makeCmpOpOverload(cmpOpScalarEQFn, a, b, false /* NullableArgs */)
+func makeEqFn(a, b *types.T, v Volatility) *CmpOp {
+	return makeCmpOpOverload(cmpOpScalarEQFn, a, b, false /* NullableArgs */, v)
 }
-func makeLtFn(a, b *types.T) *CmpOp {
-	return makeCmpOpOverload(cmpOpScalarLTFn, a, b, false /* NullableArgs */)
+func makeLtFn(a, b *types.T, v Volatility) *CmpOp {
+	return makeCmpOpOverload(cmpOpScalarLTFn, a, b, false /* NullableArgs */, v)
 }
-func makeLeFn(a, b *types.T) *CmpOp {
-	return makeCmpOpOverload(cmpOpScalarLEFn, a, b, false /* NullableArgs */)
+func makeLeFn(a, b *types.T, v Volatility) *CmpOp {
+	return makeCmpOpOverload(cmpOpScalarLEFn, a, b, false /* NullableArgs */, v)
 }
-func makeIsFn(a, b *types.T) *CmpOp {
-	return makeCmpOpOverload(cmpOpScalarIsFn, a, b, true /* NullableArgs */)
+func makeIsFn(a, b *types.T, v Volatility) *CmpOp {
+	return makeCmpOpOverload(cmpOpScalarIsFn, a, b, true /* NullableArgs */, v)
 }
 
 // CmpOps contains the comparison operations indexed by operation type.
 var CmpOps = cmpOpFixups(map[ComparisonOperator]cmpOpOverload{
 	EQ: {
 		// Single-type comparisons.
-		makeEqFn(types.AnyEnum, types.AnyEnum),
-		makeEqFn(types.Bool, types.Bool),
-		makeEqFn(types.Bytes, types.Bytes),
-		makeEqFn(types.Date, types.Date),
-		makeEqFn(types.Decimal, types.Decimal),
-		makeEqFn(types.AnyCollatedString, types.AnyCollatedString),
-		makeEqFn(types.Float, types.Float),
-		makeEqFn(types.Geography, types.Geography),
-		makeEqFn(types.Geometry, types.Geometry),
-		makeEqFn(types.INet, types.INet),
-		makeEqFn(types.Int, types.Int),
-		makeEqFn(types.Interval, types.Interval),
-		makeEqFn(types.Jsonb, types.Jsonb),
-		makeEqFn(types.Oid, types.Oid),
-		makeEqFn(types.String, types.String),
-		makeEqFn(types.Time, types.Time),
-		makeEqFn(types.TimeTZ, types.TimeTZ),
-		makeEqFn(types.Timestamp, types.Timestamp),
-		makeEqFn(types.TimestampTZ, types.TimestampTZ),
-		makeEqFn(types.Uuid, types.Uuid),
-		makeEqFn(types.VarBit, types.VarBit),
+		makeEqFn(types.AnyEnum, types.AnyEnum, VolatilityImmutable),
+		makeEqFn(types.Bool, types.Bool, VolatilityLeakProof),
+		makeEqFn(types.Bytes, types.Bytes, VolatilityLeakProof),
+		makeEqFn(types.Date, types.Date, VolatilityLeakProof),
+		makeEqFn(types.Decimal, types.Decimal, VolatilityImmutable),
+		makeEqFn(types.AnyCollatedString, types.AnyCollatedString, VolatilityLeakProof),
```

Comment:
this one doesn't feel leak proof if the collated strings don't have the same encoding, am i wrong?

Reply:
psql:
```

otan=# select 'a' collate "C" = 'b' collate "fr_FR";
ERROR:  collation mismatch between explicit collations "C" and "fr_FR"
LINE 1: select 'a' collate "C" = 'b' collate "fr_FR";
```

crdb:
```
root@127.0.0.1:63654/movr> select 'a' collate "en_AU" = 'b' collate "fr_FR";
ERROR: unsupported comparison operator: <collatedstring{en_AU}> = <collatedstring{fr_FR}>
SQLSTATE: 22023
```

guess we don't even resolve that comparator at overload resolution so in a way it's leakproof once it reaches that point?